File: app/src/main/java/com/jlxc/amapvoiceassistant/MainActivity.java
package com.jlxc.amapvoiceassistant;

import android.Manifest;
import android.app.Activity;
import android.content.ActivityNotFoundException;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.graphics.Color;
import android.net.Uri;
import android.os.Bundle;
import android.speech.RecognitionListener;
import android.speech.RecognizerIntent;
import android.speech.SpeechRecognizer;
import android.text.InputType;
import android.view.Gravity;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.Toast;

import java.util.ArrayList;
import java.util.Locale;

public class MainActivity extends Activity {
    private static final int REQ_RECORD_AUDIO = 1001;

    private SpeechRecognizer speechRecognizer;
    private TextView statusText;
    private TextView resultText;
    private EditText manualInput;
    private Button voiceButton;
    private Button navigateButton;

    private boolean isListening = false;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        buildUi();
        ensureAudioPermission();
        initSpeechRecognizer();
    }

    private void buildUi() {
        LinearLayout root = new LinearLayout(this);
        root.setOrientation(LinearLayout.VERTICAL);
        root.setGravity(Gravity.CENTER);
        root.setPadding(48, 32, 48, 32);
        root.setBackgroundColor(Color.rgb(16, 24, 32));

        TextView title = new TextView(this);
        title.setText("高德车机语音助手");
        title.setTextColor(Color.WHITE);
        title.setTextSize(34);
        title.setGravity(Gravity.CENTER);
        root.addView(title, new LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.MATCH_PARENT,
                LinearLayout.LayoutParams.WRAP_CONTENT
        ));

        statusText = new TextView(this);
        statusText.setText("说：导航到萍乡北站 / 我要去润达国际");
        statusText.setTextColor(Color.rgb(160, 220, 220));
        statusText.setTextSize(20);
        statusText.setGravity(Gravity.CENTER);
        statusText.setPadding(0, 16, 0, 12);
        root.addView(statusText);

        resultText = new TextView(this);
        resultText.setText("识别结果会显示在这里");
        resultText.setTextColor(Color.WHITE);
        resultText.setTextSize(24);
        resultText.setGravity(Gravity.CENTER);
        resultText.setPadding(0, 10, 0, 18);
        root.addView(resultText, new LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.MATCH_PARENT,
                LinearLayout.LayoutParams.WRAP_CONTENT
        ));

        manualInput = new EditText(this);
        manualInput.setHint("也可以手动输入目的地，例如：萍乡北站");
        manualInput.setSingleLine(true);
        manualInput.setTextColor(Color.WHITE);
        manualInput.setHintTextColor(Color.rgb(135, 155, 165));
        manualInput.setTextSize(22);
        manualInput.setInputType(InputType.TYPE_CLASS_TEXT);
        manualInput.setPadding(20, 10, 20, 10);
        root.addView(manualInput, new LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.MATCH_PARENT,
                LinearLayout.LayoutParams.WRAP_CONTENT
        ));

        LinearLayout buttons = new LinearLayout(this);
        buttons.setOrientation(LinearLayout.HORIZONTAL);
        buttons.setGravity(Gravity.CENTER);
        buttons.setPadding(0, 22, 0, 0);

        voiceButton = new Button(this);
        voiceButton.setText("开始语音");
        voiceButton.setTextSize(22);
        voiceButton.setOnClickListener(v -> startListening());
        buttons.addView(voiceButton, new LinearLayout.LayoutParams(0, 72, 1));

        navigateButton = new Button(this);
        navigateButton.setText("导航");
        navigateButton.setTextSize(22);
        navigateButton.setOnClickListener(v -> {
            String text = manualInput.getText().toString().trim();
            if (text.length() == 0) {
                showStatus("请先输入或语音说出目的地");
                return;
            }
            navigateToKeyword(cleanDestination(text));
        });
        LinearLayout.LayoutParams navParams = new LinearLayout.LayoutParams(0, 72, 1);
        navParams.setMargins(20, 0, 0, 0);
        buttons.addView(navigateButton, navParams);

        root.addView(buttons, new LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.MATCH_PARENT,
                LinearLayout.LayoutParams.WRAP_CONTENT
        ));

        TextView note = new TextView(this);
        note.setText("当前版本使用 androidauto://keywordNavi?keywords=目的地 调起高德车机版。");
        note.setTextColor(Color.rgb(130, 150, 160));
        note.setTextSize(16);
        note.setGravity(Gravity.CENTER);
        note.setPadding(0, 22, 0, 0);
        root.addView(note);

        setContentView(root);
    }

    private void ensureAudioPermission() {
        if (checkSelfPermission(Manifest.permission.RECORD_AUDIO) != PackageManager.PERMISSION_GRANTED) {
            requestPermissions(new String[]{Manifest.permission.RECORD_AUDIO}, REQ_RECORD_AUDIO);
        }
    }

    private void initSpeechRecognizer() {
        if (!SpeechRecognizer.isRecognitionAvailable(this)) {
            showStatus("当前系统没有可用语音识别服务，可以先用手动输入测试导航");
            return;
        }

        speechRecognizer = SpeechRecognizer.createSpeechRecognizer(this);
        speechRecognizer.setRecognitionListener(new RecognitionListener() {
            @Override public void onReadyForSpeech(Bundle params) {
                isListening = true;
                voiceButton.setText("正在听...");
                showStatus("请说目的地，例如：导航到萍乡北站");
            }

            @Override public void onBeginningOfSpeech() {
                showStatus("正在识别...");
            }

            @Override public void onRmsChanged(float rmsdB) {}

            @Override public void onBufferReceived(byte[] buffer) {}

            @Override public void onEndOfSpeech() {
                isListening = false;
                voiceButton.setText("开始语音");
                showStatus("识别结束，正在处理...");
            }

            @Override public void onError(int error) {
                isListening = false;
                voiceButton.setText("开始语音");
                showStatus("语音识别失败，错误码：" + error + "。可以手动输入目的地测试。");
            }

            @Override public void onResults(Bundle results) {
                isListening = false;
                voiceButton.setText("开始语音");

                ArrayList<String> list = results.getStringArrayList(SpeechRecognizer.RESULTS_RECOGNITION);
                if (list == null || list.isEmpty()) {
                    showStatus("没有识别到内容");
                    return;
                }

                String recognized = list.get(0);
                resultText.setText("识别：" + recognized);

                String destination = cleanDestination(recognized);
                manualInput.setText(destination);

                if (destination.length() == 0) {
                    showStatus("没有提取到目的地，请换一种说法");
                    return;
                }

                showStatus("准备导航到：" + destination);
                navigateToKeyword(destination);
            }

            @Override public void onPartialResults(Bundle partialResults) {
                ArrayList<String> list = partialResults.getStringArrayList(SpeechRecognizer.RESULTS_RECOGNITION);
                if (list != null && !list.isEmpty()) {
                    resultText.setText("正在听：" + list.get(0));
                }
            }

            @Override public void onEvent(int eventType, Bundle params) {}
        });
    }

    private void startListening() {
        if (checkSelfPermission(Manifest.permission.RECORD_AUDIO) != PackageManager.PERMISSION_GRANTED) {
            ensureAudioPermission();
            return;
        }

        if (speechRecognizer == null) {
            initSpeechRecognizer();
            if (speechRecognizer == null) {
                showStatus("当前系统没有可用语音识别服务");
                return;
            }
        }

        if (isListening) {
            speechRecognizer.stopListening();
            return;
        }

        Intent intent = new Intent(RecognizerIntent.ACTION_RECOGNIZE_SPEECH);
        intent.putExtra(RecognizerIntent.EXTRA_LANGUAGE, Locale.CHINA.toString());
        intent.putExtra(RecognizerIntent.EXTRA_LANGUAGE_PREFERENCE, "zh-CN");
        intent.putExtra(RecognizerIntent.EXTRA_PARTIAL_RESULTS, true);
        intent.putExtra(RecognizerIntent.EXTRA_MAX_RESULTS, 3);
        intent.putExtra(RecognizerIntent.EXTRA_PROMPT, "请说目的地");

        speechRecognizer.startListening(intent);
    }

    private String cleanDestination(String raw) {
        if (raw == null) return "";

        String s = raw.trim();

        String[] removeWords = new String[]{
                "小初音", "初音", "高德地图", "高德",
                "帮我", "请帮我", "麻烦",
                "导航到", "导航去", "导航", "带我去", "我要去", "我想去",
                "开车去", "去一下", "去"
        };

        for (String w : removeWords) {
            s = s.replace(w, "");
        }

        s = s.replace("。", "")
                .replace("，", "")
                .replace(",", "")
                .replace(".", "")
                .replace("！", "")
                .replace("!", "")
                .replace("？", "")
                .replace("?", "")
                .replace(" ", "")
                .trim();

        return s;
    }

    private void navigateToKeyword(String destination) {
        if (destination == null || destination.trim().length() == 0) {
            showStatus("目的地为空");
            return;
        }

        String appName = Uri.encode("高德地图");
        String keyword = Uri.encode(destination.trim());

        String uri = "androidauto://keywordNavi"
                + "?sourceApplication=" + appName
                + "&keywords=" + keyword
                + "&style=2";

        Intent intent = new Intent(Intent.ACTION_VIEW);
        intent.addCategory(Intent.CATEGORY_DEFAULT);
        intent.setData(Uri.parse(uri));

        try {
            startActivity(intent);
            showStatus("已发送到高德车机版：" + destination);
        } catch (ActivityNotFoundException e) {
            showStatus("没有找到可处理 androidauto:// 的高德车机版");
            Toast.makeText(this, "未找到高德车机版", Toast.LENGTH_LONG).show();
        } catch (Exception e) {
            showStatus("调起高德失败：" + e.getMessage());
        }
    }

    private void showStatus(String text) {
        statusText.setText(text);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        if (speechRecognizer != null) {
            speechRecognizer.destroy();
            speechRecognizer = null;
        }
    }
}

File: app/src/main/java/com/jlxc/amapvoiceassistant/MainActivity.kt
package com.jlxc.amapvoiceassistant

import android.Manifest
import android.app.Activity
import android.content.ActivityNotFoundException
import android.content.Intent
import android.content.pm.PackageManager
import android.graphics.Color
import android.media.AudioFormat
import android.media.AudioRecord
import android.media.MediaRecorder
import android.net.Uri
import android.os.Bundle
import android.text.InputType
import android.util.Log
import android.view.Gravity
import android.widget.Button
import android.widget.EditText
import android.widget.LinearLayout
import android.widget.TextView
import android.widget.Toast
import com.k2fsa.sherpa.onnx.OnlineRecognizer
import com.k2fsa.sherpa.onnx.OnlineRecognizerConfig
import com.k2fsa.sherpa.onnx.getEndpointConfig
import com.k2fsa.sherpa.onnx.getFeatureConfig
import com.k2fsa.sherpa.onnx.getModelConfig
import kotlin.concurrent.thread

private const val TAG = "AMapVoiceAssistant"
private const val REQ_RECORD_AUDIO = 200

class MainActivity : Activity() {
    private lateinit var titleText: TextView
    private lateinit var statusText: TextView
    private lateinit var resultText: TextView
    private lateinit var manualInput: EditText
    private lateinit var recordButton: Button
    private lateinit var navButton: Button

    private var recognizer: OnlineRecognizer? = null
    private var audioRecord: AudioRecord? = null
    private var recordingThread: Thread? = null

    private val sampleRateInHz = 16000
    private val channelConfig = AudioFormat.CHANNEL_IN_MONO
    private val audioFormat = AudioFormat.ENCODING_PCM_16BIT

    @Volatile
    private var isRecording = false

    @Volatile
    private var latestText = ""

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        buildUi()
        ensureAudioPermission()

        thread(start = true) {
            initOfflineRecognizer()
        }
    }

    private fun buildUi() {
        val root = LinearLayout(this).apply {
            orientation = LinearLayout.VERTICAL
            gravity = Gravity.CENTER
            setPadding(48, 28, 48, 28)
            setBackgroundColor(Color.rgb(16, 24, 32))
        }

        titleText = TextView(this).apply {
            text = "离线语音导航助手"
            setTextColor(Color.WHITE)
            textSize = 34f
            gravity = Gravity.CENTER
        }
        root.addView(titleText, LinearLayout.LayoutParams.MATCH_PARENT, LinearLayout.LayoutParams.WRAP_CONTENT)

        statusText = TextView(this).apply {
            text = "正在初始化离线识别模型，请稍等..."
            setTextColor(Color.rgb(160, 230, 225))
            textSize = 20f
            gravity = Gravity.CENTER
            setPadding(0, 14, 0, 10)
        }
        root.addView(statusText, LinearLayout.LayoutParams.MATCH_PARENT, LinearLayout.LayoutParams.WRAP_CONTENT)

        resultText = TextView(this).apply {
            text = "可以说：导航到萍乡北站 / 我要去润达国际"
            setTextColor(Color.WHITE)
            textSize = 24f
            gravity = Gravity.CENTER
            setPadding(0, 8, 0, 16)
        }
        root.addView(resultText, LinearLayout.LayoutParams.MATCH_PARENT, LinearLayout.LayoutParams.WRAP_CONTENT)

        manualInput = EditText(this).apply {
            hint = "也可以手动输入目的地"
            setSingleLine(true)
            inputType = InputType.TYPE_CLASS_TEXT
            textSize = 22f
            setTextColor(Color.WHITE)
            setHintTextColor(Color.rgb(135, 155, 165))
            setPadding(20, 10, 20, 10)
        }
        root.addView(manualInput, LinearLayout.LayoutParams.MATCH_PARENT, LinearLayout.LayoutParams.WRAP_CONTENT)

        val row = LinearLayout(this).apply {
            orientation = LinearLayout.HORIZONTAL
            gravity = Gravity.CENTER
            setPadding(0, 22, 0, 0)
        }

        recordButton = Button(this).apply {
            text = "开始离线识别"
            textSize = 22f
            isEnabled = false
            setOnClickListener {
                if (isRecording) stopRecordingAndNavigate() else startRecording()
            }
        }
        row.addView(recordButton, LinearLayout.LayoutParams(0, 72, 1f))

        navButton = Button(this).apply {
            text = "手动导航"
            textSize = 22f
            setOnClickListener {
                val dest = cleanDestination(manualInput.text.toString())
                if (dest.isBlank()) {
                    showStatus("请先输入目的地")
                } else {
                    navigateToKeyword(dest)
                }
            }
        }
        val navParams = LinearLayout.LayoutParams(0, 72, 1f)
        navParams.setMargins(20, 0, 0, 0)
        row.addView(navButton, navParams)

        root.addView(row, LinearLayout.LayoutParams.MATCH_PARENT, LinearLayout.LayoutParams.WRAP_CONTENT)

        val note = TextView(this).apply {
            text = "本版使用 sherpa-onnx Paraformer 中英双语流式模型，本地离线识别。"
            setTextColor(Color.rgb(130, 150, 160))
            textSize = 16f
            gravity = Gravity.CENTER
            setPadding(0, 22, 0, 0)
        }
        root.addView(note, LinearLayout.LayoutParams.MATCH_PARENT, LinearLayout.LayoutParams.WRAP_CONTENT)

        setContentView(root)
    }

    private fun initOfflineRecognizer() {
        try {
            showStatusOnUi("正在加载 Paraformer 离线模型...")

            val modelType = 5 // sherpa-onnx-streaming-paraformer-bilingual-zh-en
            val modelConfig = getModelConfig(modelType)
                ?: throw IllegalStateException("无法获取 sherpa-onnx model config type=$modelType")

            val config = OnlineRecognizerConfig(
                featConfig = getFeatureConfig(sampleRate = sampleRateInHz, featureDim = 80),
                modelConfig = modelConfig,
                endpointConfig = getEndpointConfig(),
                enableEndpoint = true,
            )

            recognizer = OnlineRecognizer(
                assetManager = application.assets,
                config = config,
            )

            runOnUiThread {
                recordButton.isEnabled = true
                showStatus("模型加载完成。点击按钮后说目的地，再点击停止。")
            }
        } catch (e: Throwable) {
            Log.e(TAG, "initOfflineRecognizer failed", e)
            runOnUiThread {
                recordButton.isEnabled = false
                showStatus("离线模型加载失败：" + (e.message ?: e.javaClass.simpleName))
                Toast.makeText(this, "模型加载失败，请查看 logcat", Toast.LENGTH_LONG).show()
            }
        }
    }

    private fun ensureAudioPermission() {
        if (checkSelfPermission(Manifest.permission.RECORD_AUDIO) != PackageManager.PERMISSION_GRANTED) {
            requestPermissions(arrayOf(Manifest.permission.RECORD_AUDIO), REQ_RECORD_AUDIO)
        }
    }

    private fun startRecording() {
        if (checkSelfPermission(Manifest.permission.RECORD_AUDIO) != PackageManager.PERMISSION_GRANTED) {
            ensureAudioPermission()
            return
        }

        val rec = recognizer
        if (rec == null) {
            showStatus("模型还没加载完成")
            return
        }

        if (!initMicrophone()) {
            showStatus("麦克风初始化失败")
            return
        }

        latestText = ""
        manualInput.setText("")
        resultText.text = "正在听..."
        showStatus("正在离线识别，说完后再点一次停止")
        recordButton.text = "停止并导航"

        isRecording = true
        audioRecord?.startRecording()

        recordingThread = thread(start = true) {
            processSamples(rec)
        }
    }

    private fun stopRecordingAndNavigate() {
        isRecording = false

        try {
            audioRecord?.stop()
            audioRecord?.release()
        } catch (_: Throwable) {
        }
        audioRecord = null
        recordButton.text = "开始离线识别"

        val dest = cleanDestination(latestText)
        runOnUiThread {
            manualInput.setText(dest)
            if (dest.isBlank()) {
                showStatus("没有提取到目的地，可以手动输入")
            } else {
                showStatus("识别目的地：$dest，正在调起高德")
                navigateToKeyword(dest)
            }
        }
    }

    private fun processSamples(recognizer: OnlineRecognizer) {
        val stream = recognizer.createStream()
        val intervalSeconds = 0.1
        val bufferSize = (intervalSeconds * sampleRateInHz).toInt()
        val buffer = ShortArray(bufferSize)

        try {
            while (isRecording) {
                val ret = audioRecord?.read(buffer, 0, buffer.size) ?: 0
                if (ret > 0) {
                    val samples = FloatArray(ret) { i -> buffer[i] / 32768.0f }
                    stream.acceptWaveform(samples, sampleRate = sampleRateInHz)

                    while (recognizer.isReady(stream)) {
                        recognizer.decode(stream)
                    }

                    val isEndpoint = recognizer.isEndpoint(stream)
                    var text = recognizer.getResult(stream).text

                    // Paraformer 流式模型在端点处补一点尾部静音，避免最后一个字漏识别
                    if (isEndpoint && recognizer.config.modelConfig.paraformer.encoder.isNotBlank()) {
                        val tailPaddings = FloatArray((0.8 * sampleRateInHz).toInt())
                        stream.acceptWaveform(tailPaddings, sampleRate = sampleRateInHz)
                        while (recognizer.isReady(stream)) {
                            recognizer.decode(stream)
                        }
                        text = recognizer.getResult(stream).text
                    }

                    if (text.isNotBlank()) {
                        latestText = text
                        runOnUiThread {
                            resultText.text = "识别：$text"
                        }
                    }

                    if (isEndpoint) {
                        recognizer.reset(stream)
                    }
                }
            }
        } catch (e: Throwable) {
            Log.e(TAG, "processSamples failed", e)
            runOnUiThread {
                showStatus("识别过程出错：" + (e.message ?: e.javaClass.simpleName))
            }
        } finally {
            stream.release()
        }
    }

    private fun initMicrophone(): Boolean {
        if (checkSelfPermission(Manifest.permission.RECORD_AUDIO) != PackageManager.PERMISSION_GRANTED) {
            ensureAudioPermission()
            return false
        }

        val minBytes = AudioRecord.getMinBufferSize(sampleRateInHz, channelConfig, audioFormat)
        if (minBytes <= 0) return false

        audioRecord = AudioRecord(
            MediaRecorder.AudioSource.MIC,
            sampleRateInHz,
            channelConfig,
            audioFormat,
            minBytes * 2
        )

        return audioRecord?.state == AudioRecord.STATE_INITIALIZED
    }

    private fun cleanDestination(raw: String?): String {
        if (raw == null) return ""

        var s = raw.trim()

        val removeWords = arrayOf(
            "小初音", "初音", "高德地图", "高德",
            "请帮我", "帮我", "麻烦",
            "导航到", "导航去", "导航", "带我去", "我要去", "我想去",
            "开车去", "去一下", "我去", "去"
        )

        for (word in removeWords) {
            s = s.replace(word, "")
        }

        return s.replace("。", "")
            .replace("，", "")
            .replace(",", "")
            .replace(".", "")
            .replace("！", "")
            .replace("!", "")
            .replace("？", "")
            .replace("?", "")
            .replace(" ", "")
            .trim()
    }

    private fun navigateToKeyword(destination: String) {
        val dest = destination.trim()
        if (dest.isBlank()) {
            showStatus("目的地为空")
            return
        }

        val uri = "androidauto://keywordNavi" +
            "?sourceApplication=${Uri.encode("高德地图")}" +
            "&keywords=${Uri.encode(dest)}" +
            "&style=2"

        val intent = Intent(Intent.ACTION_VIEW).apply {
            addCategory(Intent.CATEGORY_DEFAULT)
            data = Uri.parse(uri)
        }

        try {
            startActivity(intent)
            showStatus("已发送到高德车机版：$dest")
        } catch (e: ActivityNotFoundException) {
            showStatus("没有找到可处理 androidauto:// 的高德车机版")
            Toast.makeText(this, "未找到高德车机版", Toast.LENGTH_LONG).show()
        } catch (e: Throwable) {
            showStatus("调起高德失败：" + (e.message ?: e.javaClass.simpleName))
        }
    }

    private fun showStatus(text: String) {
        statusText.text = text
    }

    private fun showStatusOnUi(text: String) {
        runOnUiThread { showStatus(text) }
    }

    override fun onDestroy() {
        isRecording = false
        try {
            audioRecord?.stop()
            audioRecord?.release()
        } catch (_: Throwable) {
        }
        audioRecord = null

        try {
            recognizer?.release()
        } catch (_: Throwable) {
        }
        recognizer = null

        super.onDestroy()
    }
}
